### PR TITLE
docs: add lua 5.1 as part of the neovim documentation

### PR DESCRIPTION
**Pull request by dundargoc**
I think it'd be nice to have in core but I'm open to thoughts.

Snatched this from https://github.com/bfredl/luarefvim

### CODE REVIEW

**Comment by lewis6991** [comment]
This keeps with the idiom that all documentation for Neovim is available offlin via the `:help` system.

If we do this, we should also add luv/libuv.

**Comment by clason** [comment]
We might want to massage this somewhat, though -- as far as I can tell, this file uses additional syntax that requires(?) the syntax and plugin files distributed with it.

> If we do this, we should also add luv/libuv.

That is a good idea, but someone would have to create it first. Might as well turn it into a custom `vim.loop` documentation, then.

File: runtime/doc/lua51refvim.txt
Comment (bfredl) [inline runtime/doc/lua51refvim.txt:26]: we probably change the prefix to something less keysmashy, maybe `luaref`, which also can be the filename. ("LuaRef" means something else in the C code base, but this term is not leaked to public docs, so there is no confusion)

File: runtime/doc/help.txt
Comment (bfredl) [inline runtime/doc/help.txt:188]: just Lua not "lua 5.1" in most places (except when in a direct quote like line 14).

**Comment by bfredl** [comment]
Either needs to add these syntax classes or strip them out (likely some could folded into the standard help syntax groups)

```
" add three syntax classes: bold, emph (italic) and code -- similarly to html
"
" TODO: match luarefvim filename?
syn match helpIgnoreBold     "#[a-zA-Z0-9&()\`\'\"\-\+\*=\[\]\{\}\.,;: ]\+#" contains=helpMatchBold
syn match helpMatchBold      "[a-zA-Z0-9&()\`\'\"\-\+\*=\[\]\{\}\.,;: ]\+"   contained
syn match helpIgnoreEmph     "@[a-zA-Z0-9&()\`\'\"\-\+\*=\[\]\{\}\.,;: ]\+@" contains=helpMatchEmph
syn match helpMatchEmph      "[a-zA-Z0-9&()\`\'\"\-\+\*=\[\]\{\}\.,;: ]\+"   contained
" this match is the same as in CRefVim's help.vim (that is, uses $$):
" the idea is to keep some degree of portability.
" TODO: convert code blocks to standard < > notation
syn match helpIgnoreCode     "\$[a-zA-Z0-9@\\\*/\._=()\-+%<>&\^|!~\?:,\[\];{}#\`\'\" ]\+\$" contains=helpMatchCode
syn match helpMatchCode      "[a-zA-Z0-9@\\\*/\._=()\-+%<>&\^|!~\?:,\[\];{}#\`\'\" ]\+"   contained
```

**Comment by lewis6991** [comment]
> That is a good idea, but someone would have to create it first. Might as well turn it into a custom vim.loop documentation, then.

A plugin already exists for it. On mobile but will update later.

**Comment by dundargoc** [comment]
> A plugin already exists for it: [nanotee/luv-vimdocs](https://github.com/nanotee/luv-vimdocs)

hippity hoppity, your plugin is now my property :)

I've added it, thanks for the tip.

**Comment by dundargoc** [comment]
I should mention that I haven't done any sort of proofreading of the documents yet, I wanted to get a feeling for the overall reception before putting in mountains of work :)

File: runtime/doc/luarefvim.txt
Comment (bfredl) [inline runtime/doc/luarefvim.txt:5572]: entire C.1 and C.2 are irrelevant when shipped in runtime, could just be deleted

**Comment by bfredl** [comment]
https://github.com/wsdjeg/luarefvim/blob/master/plugin/help.vim the last six  `hi link` here would also be needed, though ideally we will completely eliminate the special $ code blocks at least.

File: runtime/doc/luarefvim.txt
Comment (clason) [inline runtime/doc/luarefvim.txt:5560]: That's not correct anymore?

File: runtime/doc/luarefvim.txt
Comment (clason) [inline runtime/doc/luarefvim.txt:5403]: irrelevant?

File: runtime/doc/help.txt
Comment (clason) [inline runtime/doc/help.txt:188]: I'd use just `luaref.txt` (or `luareference.txt`, for disambiguation) -- the `vim` part is slightly misleading since the documentation is _not_ Vim-related (only the format is).

**Comment by clason** [comment]
Do we want to (and can we) move the copyright to `LICENSE.txt`? There's a "external library" section there which could cover this. (@justinmk)

**Comment by clason** [comment]
Did a bit of cleanup and removed all the weird bold markup.

**Comment by zeertzjq** [comment]
These are still many mentions of `luarefvim` instead of `luarefman`.

**Comment by clason** [comment]
> These are still many mentions of `luarefvim` instead of `luarefman`.

Yes, I wanted to wait for feedback on the renaming before making (and having to undo) a lot of changes.

**Comment by lewis6991** [comment]
I've done a significantly large reformat of `luarefman.txt` to be consistent with `lua.txt`. 

Let me know if anyone disagrees with anything I've done, or whether they would prefer the format be done more incrementally in multiple commits.

If we are happy with it, I can do the same for libuv.

**Comment by clason** [comment]
No, better to do it right the first time. (The luv docs look fine as they are, though? @nanotee already did a great job.)

**Comment by lewis6991** [comment]
> No, better to do it right the first time. (The luv docs look fine as they are, though? @nanotee already did a great job.)

Haven't looked at them yet.

For reference, I used this lua to auto justify the tags:

```lua
local TW = 78

for i, l in ipairs(vim.api.nvim_buf_get_lines(0, 0, -1, false)) do
  if vim.endswith(l, '*') then
    local newline
    local _, nstars = l:gsub('*', '')

    -- Match padding + tags
    local suffix = l:match('%s*%*.*')
    local prefix = l:sub(1, -#suffix-1)

    if #l > TW then
      local offset = #l - TW - nstars

      -- Check we have enough padding to remove
      if vim.startswith(suffix, string.rep(' ', offset)) then
        newline = prefix..suffix:sub(offset+1)
      end
    elseif #l < TW then
      local offset = TW - #l + nstars
      newline = prefix..string.rep(' ', offset)..suffix
    end

    if newline then
      vim.api.nvim_buf_set_lines(0, i-1, i, false, {newline})
    end
  end
end
```

In case we want to tweak it further, or use this as a basis for other stuff.

**Comment by clason** [comment]
I think you removed too much from the `LUAREFVIM DOC` heading? (File is too big to comment on it...)

**Comment by clason** [comment]
Now that we are already this close to native docs, we might as well fix the final nit for me: the colon in list items like

>  - `\a` : bell

File: runtime/doc/help.txt
Comment (clason) [inline runtime/doc/help.txt:188]: Maybe we should split off `api.txt`, `lua.txt`, and this (and the luv) reference manual to a separate section ("APIs")?

**Comment by dundargoc** [comment]
Bit of a bikeshed: maybe just `luaref` instead of `luarefman`? Luarefman sounds like a bad super hero name (strong Macro Man vibes :) )

**Comment by bfredl** [comment]
just luaref is fine. `LuaRef` is completely internal and thus not a name collision in public documentation.

**Comment by dundargoc** [comment]
Took a quick look at it, looks awesome. Thanks for the work @clason and @lewis6991. I don't have any improvement ideas as you've already taken care of all of them :). I think it looks good enough to ship, but I'll let you decide.

**Review by lewis6991 — APPROVED** [review]

**Review by lewis6991 — APPROVED** [review]

**Comment by clason** [comment]
yay teamwork \o/

**Pull request merged**
